### PR TITLE
torch integration for DefaultQubit2

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
Integrating the interfaces with the new device interface. Same as #4137, but with torch!

**Description of the Change:**
- Convert tuples to lists when changing them in place, but otherwise nothing in code
- add tests

**Benefits:**
Torch works with the new device interface

**Possible Drawbacks:**
some tests feel a bit more silly, like using a lambda (as the torch docs recommend) to skip non-trainable params when computing the jacobian, and then asserting that the non-trainable param had no effect on the jacobian. but that's the beauty of torch:) had to skip a few tests for not being torch-y

### CODE REVIEW

**Review by albi3ro — COMMENTED** [review]
I'll be good to approve once those two remaining tests are fixed.

**Comment by timmysilv** [comment]
> I'll be good to approve once those two remaining tests are fixed.

done! I set the global tensor type to double instead of float (otherwise every assertion complains about this), and some other tests depended on it, so now I undo this after each test.

**Review by lillian542 — APPROVED** [review]
Looks good to me! 🚀

File: tests/interfaces/default_qubit_2_integration/test_torch_default_qubit_2.py [hunk 1]
```diff
@@ -263,38 +285,39 @@ def cost(params):
...
-        assert np.allclose(jac[3, 0], d1, atol=atol_for_shots(shots))
+        d1 = -torch.sin(x) * torch.cos(y)
+        assert torch.allclose(jac[0, 0], d1, atol=atol_for_shots(shots))  # fails for torch
+        assert torch.allclose(jac[3, 0], d1, atol=atol_for_shots(shots))
 
-        d2 = -np.cos(x) * np.sin(y)
-        assert np.allclose(jac[0, 1], d2, atol=atol_for_shots(shots))
-        assert np.allclose(jac[3, 1], d2, atol=atol_for_shots(shots))
+        d2 = -torch.cos(x) * torch.sin(y)
+        assert torch.allclose(jac[0, 1], d2, atol=atol_for_shots(shots))  # fails for torch
```
[the remark this answers is not in the document]
Comment (timmysilv) [reply tests/interfaces/default_qubit_2_integration/test_torch_default_qubit_2.py:320]: Torch puts gradient information directly on the Tensor objects, so they can't be re-used across tapes effectively. I've reproduced this code with torch vs. autograd if you're curious, but the jacobian makes it clear what went wrong.

<details>
<summary>Code hidden here:)</summary>

```python
import pennylane as qml
import numpy as np
​
import torch
from pennylane import numpy as qnp
​
device = qml.devices.experimental.DefaultQubit2()
shots = None
​
def cost(params):
    tape1 = qml.tape.QuantumScript(
        [qml.RY(params[0], 0), qml.RX(params[1], 0)],
        [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))],
        shots=shots,
    )
​
    tape2 = qml.tape.QuantumScript(
        [qml.RY(np.array(0.5), 0)],
        [qml.expval(qml.PauliZ(0))],
        shots=shots,
    )
​
    tape3 = qml.tape.QuantumScript(
        [qml.RY(params[0], 0), qml.RX(params[1], 0)],
        [qml.expval(qml.PauliZ(0))],
        shots=shots,
    )
    return qml.execute([tape1, tape2, tape3], device, gradient_fn=qml.gradients.param_shift)
​
def cost_torch(params):
    res = cost(params)
    return torch.hstack([qml.math.asarray(r, like="torch") for r in res])
​
def cost_autograd(params):
    return qml.math.hstack(cost(params))
​
params = torch.tensor([0.1, 0.2], requires_grad=True)
print("torch:\n", torch.autograd.functional.jacobian(cost_torch, params))
​
params = qnp.array([0.1, 0.2], requires_grad=True)
print("autograd:\n", qml.jacobian(cost_autograd)(params))
```
</details>

```
torch:
 tensor([[ 0.0000,  0.0000],
        [ 0.0000,  0.0000],
        [ 0.0000,  0.0000],
        [-0.0978, -0.1977]])
autograd:
 [[-0.0978434  -0.19767681]
 [ 0.          0.        ]
 [ 0.          0.        ]
 [-0.0978434  -0.19767681]]
```

[the remark this answers is not in the document]
Comment (lillian542) [reply tests/interfaces/default_qubit_2_integration/test_torch_default_qubit_2.py:320]: Thanks!

**Comment by albi3ro** [comment]
[sc-37946]
